File: 搜索/78.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> one = new ArrayList<>();
        dfs(nums,0,one,res);
        return res;
    }

    public void dfs(int[] nums, int start, List<Integer> one, List<List<Integer>> res)
    {
        if(start == nums.length)
        {
            List<Integer> t = new ArrayList<>(one);
            res.add(t);
            return;
        }
        one.add(nums[start]);
        dfs(nums, start+1, one, res);
        one.remove(one.size()-1);
        dfs(nums, start+1, one, res);
    }
}

File: 搜索/78.go
func subsets(nums []int) [][]int {
    ans := make([][]int, 0)
    one := make([]int, 0)
    dfs(nums, 0, &ans, &one)
    return ans
}

func dfs(nums []int, start int, ans *[][]int, one *[]int) {
    t := make([]int, len(*one))
    copy(t, *one)
    *ans = append(*ans, t)
    for i := start; i < len(nums); i++ {
        *one = append(*one, nums[i])
        dfs(nums, i + 1, ans, one)
        *one = (*one)[:len(*one) - 1]
    }
}
